pesto/optimize/optimizer.py

```diff
@@ -0,0 +1,234 @@
+import scipy.optimize
+import re
+import sys
+import abc
+
+
+class OptimizerResult(dict):
+    """
+    The result of an optimizer run. Used as a standardized return value to
+    map from the individual result objects returned by the employed
+    optimizers to the format understood by pesto.
+
+    Can be used like a dict.
+
+    Attributes
+    ----------
+
+    x: ndarray
+        The best found parameters.
+
+    fval: float
+        The best found function value, fun(x).
+
+    grad, hess: ndarray
+        The gradient and Hessian at x.
+
+    n_fval: int
+        Number of function evaluations.
+
+    n_grad: int
+        Number of gradient evaluations.
+
+    n_hess: int
+        Number of Hessian evaluations.
+
+    exitflag: int
+        The exitflag of the optimizer.
+
+    message: str
+        Textual comment on the optimization result.
+
+    Any field not supported by the optimizer is filled with None.
```

Comment:
true, there should at least be fields for that, even if not all optimizers support that. also a field for time seems to be missing, and we need to allow the user to specify whether or not to store certain values, in particular hessians.

Reply:
Yes. Regarding time: we should already think about a way the cost function can provide its own computation time, which would be very helpful in case of parallelization in there. I thought about making the objective function return a `dict` instead of the current `tuple`. Not sure if it is my favourite solution, though. It would allow for passing arbitrary values to be stored in the optimization result, which could become handy in the future. Also it would be more descriptive than using the numeric indices.

Reply:
good idea, vote for dict. a problem might be that we then need to have another wrapper for scipy. this wrapper might be located in the Optimizer routine and extract the needed values / tuples and pass them on to the local optimizer.

Reply:
@dweindl Thought about dict again? Would make the thing more customizable, but sth like fval = objective(x) would become fval = objective(x)['fval']. The time is currently taken in the Optimizer class, see https://github.com/ICB-DCM/pyPESTO/blob/feature_optim_results/pesto/optimize/optimizer.py#L87.

Reply:
@yannikschaelte No further ideas yet. `fval = objective(x)['fval']` is indeed unfavorable. Could provide a dict as output argument, but that is not really the python way. Returning (fva, dict) is as ugly as the previous example.